flink-connectors/flink-hbase/src/main/java/org/apache/flink/addons/hbase/HBaseUpsertTableSink.java

```diff
@@ -0,0 +1,130 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.addons.hbase;
+
+import org.apache.flink.annotation.VisibleForTesting;
+import org.apache.flink.api.common.typeinfo.TypeInformation;
+import org.apache.flink.api.java.tuple.Tuple2;
+import org.apache.flink.streaming.api.datastream.DataStream;
+import org.apache.flink.streaming.api.datastream.DataStreamSink;
+import org.apache.flink.table.api.TableSchema;
+import org.apache.flink.table.api.ValidationException;
+import org.apache.flink.table.sinks.TableSink;
+import org.apache.flink.table.sinks.UpsertStreamTableSink;
+import org.apache.flink.table.utils.TableConnectorUtils;
+import org.apache.flink.types.Row;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.hbase.HBaseConfiguration;
+import org.apache.hadoop.hbase.HConstants;
+
+import java.util.Arrays;
+
+import static org.apache.flink.util.Preconditions.checkArgument;
+
+/**
+ * An upsert {@link UpsertStreamTableSink} for HBase.
+ */
+public class HBaseUpsertTableSink implements UpsertStreamTableSink<Row> {
+
+	private final HBaseTableSchema hbaseTableSchema;
+	private final TableSchema tableSchema;
+	private final HBaseOptions hbaseOptions;
+	private final HBaseWriteOptions writeOptions;
+
+	public HBaseUpsertTableSink(
+			HBaseTableSchema hbaseTableSchema,
+			HBaseOptions hbaseOptions,
+			HBaseWriteOptions writeOptions) {
+		checkArgument(hbaseTableSchema.getRowKeyName().isPresent(), "HBaseUpsertTableSink requires rowkey is set.");
+		this.hbaseTableSchema = hbaseTableSchema;
+		this.tableSchema = hbaseTableSchema.convertsToTableSchema();
+		this.hbaseOptions = hbaseOptions;
+		this.writeOptions = writeOptions;
+	}
+
+	@Override
+	public void setKeyFields(String[] keys) {
+		// hbase always upsert on rowkey, ignore query keys.
```

Comment:
Check to row keys?

Reply:
I'm not sure. Because the support for key derivation is not very well, esp. fields concating which is used heavily in hbase sink.

Reply:
Can you add some comments in HBaseUpsertTableSink to explain it? It really confused...